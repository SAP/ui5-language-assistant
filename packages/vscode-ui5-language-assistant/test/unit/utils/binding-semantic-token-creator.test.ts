import {
  getSemanticTokens,
  getTokenType,
  CustomSemanticToken,
} from "../../../src/utils";
import {
  Config,
  ProjectName,
  ProjectType,
  TestFramework,
} from "@ui5-language-assistant/test-framework";
import { SemanticTokenTypes } from "vscode-languageserver-types";

describe("binding-semantic-token-creator", () => {
  let testFramework: TestFramework;
  let documentUri;
  const viewFilePathSegments = [
    "app",
    "manage_travels",
    "webapp",
    "ext",
    "main",
    "Main.view.xml",
  ];
  beforeAll(function () {
    const useConfig: Config = {
      projectInfo: {
        name: ProjectName.cap,
        type: ProjectType.CAP,
        npmInstall: false,
      },
    };
    testFramework = new TestFramework(useConfig);
    documentUri = testFramework.getFileUri(viewFilePathSegments);
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });
  const getContent = (binding: string) => `
<mvc:View
      xmlns="sap.m"
      xmlns:mvc="sap.ui.core.mvc">
      <Text id="test-id" text="${binding}"/>
</mvc:View>
  `;
  const getAggregationContent = (binding: string) => `
<mvc:View
      xmlns="sap.m"
      xmlns:mvc="sap.ui.core.mvc">
      <HeaderContainer id="test-id" content="${binding}"/>
</mvc:View>
  `;
  describe("getSemanticTokens", () => {
    it("get semantics token for curly", async () => {
      const content = getContent("{}");
      const result = await getSemanticTokens({ documentUri, content });
      const bracket = getTokenType(CustomSemanticToken.bracket);
      expect(result.filter((i) => i.tokenType === bracket).length).toEqual(2);
    });
    it("get semantics token for braces", async () => {
      const content = getContent("{parts: []}");
      const result = await getSemanticTokens({ documentUri, content });
      const bracket = getTokenType(CustomSemanticToken.bracket);
      expect(result.filter((i) => i.tokenType === bracket).length).toEqual(4);
    });
    it("get semantics token for curly as value", async () => {
      const content = getContent("{anyKey: {}}");
      const result = await getSemanticTokens({ documentUri, content });
      const bracket = getTokenType(CustomSemanticToken.bracket);
      expect(result.filter((i) => i.tokenType === bracket).length).toEqual(4);
    });

    it("get semantics token for string", async () => {
      const content = getContent("{path: 'some-value'}");
      const result = await getSemanticTokens({ documentUri, content });
      const tokenType = getTokenType(SemanticTokenTypes.string);
      const semanticResult = result.filter((i) => i.tokenType === tokenType);
      expect(semanticResult.length).toEqual(1);
    });
    it("get semantics token for number", async () => {
      const content = getContent("{anyKey: 123}");
      const result = await getSemanticTokens({ documentUri, content });
      const tokenType = getTokenType(SemanticTokenTypes.number);
      const semanticResult = result.filter((i) => i.tokenType === tokenType);
      expect(semanticResult.length).toEqual(1);
    });
    it("get semantics token for boolean", async () => {
      const content = getContent("{anyKey: true}");
      const result = await getSemanticTokens({ documentUri, content });
      const tokenType = getTokenType(CustomSemanticToken.boolean);
      const semanticResult = result.filter((i) => i.tokenType === tokenType);
      expect(semanticResult.length).toEqual(1);
    });
    it("get semantics token for null", async () => {
      const content = getContent("{anyKey: null}");
      const result = await getSemanticTokens({ documentUri, content });
      const tokenType = getTokenType(CustomSemanticToken.null);
      const semanticResult = result.filter((i) => i.tokenType === tokenType);
      expect(semanticResult.length).toEqual(1);
    });
    it("get semantics token for comma or colon [operator]", async () => {
      const content = getContent("{anyKey: null, anyKey02: true}");
      const result = await getSemanticTokens({ documentUri, content });
      const tokenType = getTokenType(SemanticTokenTypes.operator);
      const semanticResult = result.filter((i) => i.tokenType === tokenType);
      expect(semanticResult.length).toEqual(3);
    });
    it("test semantic tokens for property", async () => {
      const snippet = `
        {
          parts: [
            {value: 'formatWithBrackets'},
            {path: 'scholar/employeeNav/manager/displayName', targetType: 'any'},
            {
              path: 'scholar/employeeNav/manager_userID',
              type: 'sap.ui.model.odata.type.String',
              constraints: {maxLength: 255},
              formatOptions: {parseKeepsEmptyString: true}
            }
          ],
          formatter: 'sap.fe.core.formatters.ValueFormatter'
        }
      `;
      const content = getContent(snippet);
      const result = await getSemanticTokens({ documentUri, content });
      expect(result.length).toEqual(51);
      expect(result).toMatchSnapshot();
    });
    it("test semantic tokens for aggregation", async () => {
      const snippet = `
      {
        path:'/Travel', 
        parameters : {
            $filter : 'filter logic goes here',
            $orderby : 'TotalPrice desc'
        }
      }`;
      const content = getAggregationContent(snippet);
      const result = await getSemanticTokens({ documentUri, content });
      expect(result.length).toEqual(17);
      expect(result).toMatchSnapshot();
    });
    describe("no semantic token", () => {
      it("undefined root element [!ast.rootElement]", async () => {
        const result = await getSemanticTokens({
          documentUri,
          content: "<ab></ab>",
        });
        expect(result.length).toEqual(0);
      });
      it("not ui5 node", async () => {
        const content = `
<mvc:View
      xmlns="sap.m"
      xmlns:mvc="sap.ui.core.mvc">
      <Text id="test-id" ABC="{path: 'some-value'}"/>
</mvc:View>
  `;
        const result = await getSemanticTokens({
          documentUri,
          content,
        });
        expect(result.length).toEqual(0);
      });
      it("binding expression", async () => {
        const content = getContent("{:= some-binding-expression }");
        const result = await getSemanticTokens({ documentUri, content });
        expect(result.length).toEqual(0);
      });
      it("not allowed binding", async () => {
        const content = getContent("{path}");
        const result = await getSemanticTokens({ documentUri, content });
        expect(result.length).toEqual(0);
      });
      it("not xml view", async () => {
        const content = getContent("{path: 'someValue'}");
        const result = await getSemanticTokens({
          documentUri: "not-a-view-file",
          content,
        });
        expect(result.length).toEqual(0);
      });
    });
  });
});
