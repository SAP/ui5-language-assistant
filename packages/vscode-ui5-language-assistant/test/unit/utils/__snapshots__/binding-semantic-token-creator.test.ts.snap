// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`binding-semantic-token-creator getSemanticTokens test semantic tokens for aggregation 1`] = `
Array [
  Object {
    "char": 52,
    "length": 1,
    "line": 4,
    "tokenType": 5,
  },
  Object {
    "char": 8,
    "length": 4,
    "line": 5,
    "tokenType": 0,
  },
  Object {
    "char": 12,
    "length": 1,
    "line": 5,
    "tokenType": 3,
  },
  Object {
    "char": 13,
    "length": 9,
    "line": 5,
    "tokenType": 1,
  },
  Object {
    "char": 8,
    "length": 10,
    "line": 6,
    "tokenType": 0,
  },
  Object {
    "char": 19,
    "length": 1,
    "line": 6,
    "tokenType": 3,
  },
  Object {
    "char": 21,
    "length": 1,
    "line": 6,
    "tokenType": 5,
  },
  Object {
    "char": 12,
    "length": 7,
    "line": 7,
    "tokenType": 0,
  },
  Object {
    "char": 20,
    "length": 1,
    "line": 7,
    "tokenType": 3,
  },
  Object {
    "char": 22,
    "length": 24,
    "line": 7,
    "tokenType": 1,
  },
  Object {
    "char": 12,
    "length": 8,
    "line": 8,
    "tokenType": 0,
  },
  Object {
    "char": 21,
    "length": 1,
    "line": 8,
    "tokenType": 3,
  },
  Object {
    "char": 23,
    "length": 17,
    "line": 8,
    "tokenType": 1,
  },
  Object {
    "char": 46,
    "length": 1,
    "line": 7,
    "tokenType": 3,
  },
  Object {
    "char": 8,
    "length": 1,
    "line": 9,
    "tokenType": 5,
  },
  Object {
    "char": 22,
    "length": 1,
    "line": 5,
    "tokenType": 3,
  },
  Object {
    "char": 6,
    "length": 1,
    "line": 10,
    "tokenType": 5,
  },
]
`;

exports[`binding-semantic-token-creator getSemanticTokens test semantic tokens for property 1`] = `
Array [
  Object {
    "char": 40,
    "length": 1,
    "line": 4,
    "tokenType": 5,
  },
  Object {
    "char": 10,
    "length": 5,
    "line": 5,
    "tokenType": 0,
  },
  Object {
    "char": 15,
    "length": 1,
    "line": 5,
    "tokenType": 3,
  },
  Object {
    "char": 17,
    "length": 1,
    "line": 5,
    "tokenType": 5,
  },
  Object {
    "char": 12,
    "length": 1,
    "line": 6,
    "tokenType": 5,
  },
  Object {
    "char": 13,
    "length": 5,
    "line": 6,
    "tokenType": 0,
  },
  Object {
    "char": 18,
    "length": 1,
    "line": 6,
    "tokenType": 3,
  },
  Object {
    "char": 20,
    "length": 20,
    "line": 6,
    "tokenType": 1,
  },
  Object {
    "char": 40,
    "length": 1,
    "line": 6,
    "tokenType": 5,
  },
  Object {
    "char": 12,
    "length": 1,
    "line": 7,
    "tokenType": 5,
  },
  Object {
    "char": 13,
    "length": 4,
    "line": 7,
    "tokenType": 0,
  },
  Object {
    "char": 17,
    "length": 1,
    "line": 7,
    "tokenType": 3,
  },
  Object {
    "char": 19,
    "length": 41,
    "line": 7,
    "tokenType": 1,
  },
  Object {
    "char": 62,
    "length": 10,
    "line": 7,
    "tokenType": 0,
  },
  Object {
    "char": 72,
    "length": 1,
    "line": 7,
    "tokenType": 3,
  },
  Object {
    "char": 74,
    "length": 5,
    "line": 7,
    "tokenType": 1,
  },
  Object {
    "char": 60,
    "length": 1,
    "line": 7,
    "tokenType": 3,
  },
  Object {
    "char": 79,
    "length": 1,
    "line": 7,
    "tokenType": 5,
  },
  Object {
    "char": 12,
    "length": 1,
    "line": 8,
    "tokenType": 5,
  },
  Object {
    "char": 14,
    "length": 4,
    "line": 9,
    "tokenType": 0,
  },
  Object {
    "char": 18,
    "length": 1,
    "line": 9,
    "tokenType": 3,
  },
  Object {
    "char": 20,
    "length": 36,
    "line": 9,
    "tokenType": 1,
  },
  Object {
    "char": 14,
    "length": 4,
    "line": 10,
    "tokenType": 0,
  },
  Object {
    "char": 18,
    "length": 1,
    "line": 10,
    "tokenType": 3,
  },
  Object {
    "char": 20,
    "length": 32,
    "line": 10,
    "tokenType": 1,
  },
  Object {
    "char": 14,
    "length": 11,
    "line": 11,
    "tokenType": 0,
  },
  Object {
    "char": 25,
    "length": 1,
    "line": 11,
    "tokenType": 3,
  },
  Object {
    "char": 27,
    "length": 1,
    "line": 11,
    "tokenType": 5,
  },
  Object {
    "char": 28,
    "length": 9,
    "line": 11,
    "tokenType": 0,
  },
  Object {
    "char": 37,
    "length": 1,
    "line": 11,
    "tokenType": 3,
  },
  Object {
    "char": 39,
    "length": 3,
    "line": 11,
    "tokenType": 2,
  },
  Object {
    "char": 42,
    "length": 1,
    "line": 11,
    "tokenType": 5,
  },
  Object {
    "char": 14,
    "length": 13,
    "line": 12,
    "tokenType": 0,
  },
  Object {
    "char": 27,
    "length": 1,
    "line": 12,
    "tokenType": 3,
  },
  Object {
    "char": 29,
    "length": 1,
    "line": 12,
    "tokenType": 5,
  },
  Object {
    "char": 30,
    "length": 21,
    "line": 12,
    "tokenType": 0,
  },
  Object {
    "char": 51,
    "length": 1,
    "line": 12,
    "tokenType": 3,
  },
  Object {
    "char": 53,
    "length": 4,
    "line": 12,
    "tokenType": 4,
  },
  Object {
    "char": 57,
    "length": 1,
    "line": 12,
    "tokenType": 5,
  },
  Object {
    "char": 56,
    "length": 1,
    "line": 9,
    "tokenType": 3,
  },
  Object {
    "char": 52,
    "length": 1,
    "line": 10,
    "tokenType": 3,
  },
  Object {
    "char": 43,
    "length": 1,
    "line": 11,
    "tokenType": 3,
  },
  Object {
    "char": 12,
    "length": 1,
    "line": 13,
    "tokenType": 5,
  },
  Object {
    "char": 41,
    "length": 1,
    "line": 6,
    "tokenType": 3,
  },
  Object {
    "char": 80,
    "length": 1,
    "line": 7,
    "tokenType": 3,
  },
  Object {
    "char": 10,
    "length": 1,
    "line": 14,
    "tokenType": 5,
  },
  Object {
    "char": 10,
    "length": 9,
    "line": 15,
    "tokenType": 0,
  },
  Object {
    "char": 19,
    "length": 1,
    "line": 15,
    "tokenType": 3,
  },
  Object {
    "char": 21,
    "length": 39,
    "line": 15,
    "tokenType": 1,
  },
  Object {
    "char": 11,
    "length": 1,
    "line": 14,
    "tokenType": 3,
  },
  Object {
    "char": 8,
    "length": 1,
    "line": 16,
    "tokenType": 5,
  },
]
`;
